docs/guide/programming-tasks.md

```diff
@@ -0,0 +1,288 @@
+## Introduction into Starlark
+
+Most commonly, Cirrus tasks are declared in `.cirrus.yml` file in YAML format as documented in [*Writing Tasks*](writing-tasks.md) guide.
+
+YAML, as a language, is great for declaring simple to moderate configurations but sometimes just using a declarative language is not enough.
+One might need some conditional execution or have an easy way to generate multiple similar tasks. Most of the CIs solve this problem
+by introducing special DSL into the existing YAML. In case of Cirrus CI, we have [`only_if` keyword](writing-tasks.md#conditional-task-execution)
+for conditional execution and [`matrix` modification](writing-tasks.md#matrix-modification) for generating similar tasks.
+These optional are mostly hacks to workaround declarative nature of YAML language where in reality an imperative language
+looks like a better fit. This is why Cirrus CI allows in additional to YAML configure tasks via Starlark.
+
+Starlark language is a procedural programming language [originated from Bazel build tool](https://docs.bazel.build/versions/master/skylark/language.html)
+but ideal for embedding within any other system that want to safely allow user-defined logic. There are a few key differences which made us
+choose Starlark instead of common alternatives like JavaScript/TypeScript or WebAssembly:
+
+1. Starlark doesn't require compilation. No need to introduce full-blown compile -> deploy process for a few dozen lines of logic.
+2. Starlark script can be executed instantly on any platform. There is Starlark interpreter written in Go which integrates nicely with Cirrus CLI / Cirrus CI infrastructure.
+3. Starlark has built-in functionality for loading external modules which is ideal for config sharing. See [module loading](#module-loading) for details.
+
+## Writing Starlark scripts
+
+Let's start with a trivial `.cirrus.star` example like this:
+
+```python
+def main():
+    return [
+        {
+            "container": {
+                "image": "debian:latest",
+            },
+            "script": "make",
+        },
+    ]
+```
+
+`main` simply returns a list of task objects which will be serialized into YAML presentation like this:
+
+```yaml
+task:
+    container:
+      image: debian:latest
+    script: make
+```
+
+Then the generated YAML is appended to `.cirrus.yml` (if any) before passing the combined config into the final YAML parser.
+
+With Starlark, it's possible to generate parts of the configuration dynamically based on some external conditions: by [making an HTTP request](#http) to check the previous build status
```

Comment:
It would be nice to refactor these two examples into a bullet list, because it's rather hard to read this sentence in one pass due to it being loaded.

Reply:
See 829a048f7c79a22be75a1ab60269dd89ced082d6